### PR TITLE
Pass in commissioning delegate in init parameters

### PR DESCRIPTION
**Pull request by cecille**
#### Problem
The commissioning delegate was moved out from DeviceCommissioner in order to give vendors more control over when each commissioning stage was called. The AutoCommissioner was originally intended to be provided as an example commissioner that developers could use as the basis for their own commissioning flows. However, it is currently hooked up directly to the Commission etc. APIs, so the way to use a custom commissioner was unintuitive. This allows a default commissioner to be passed in at initialization, so it can be triggered by calls to Commission.

#### Change overview
- add init parameter to set a default commissioning delegate to control the commissining flow
- plumbs a test commissioning delegate into the python
- adds cirque tests to prove the test commissioner is used.

#### Testing
manual by chip-device-ctrl, cirque test added.

### CODE REVIEW

File: src/controller/CHIPDeviceController.cpp [hunk 1]
```diff
@@ -684,6 +684,14 @@ CHIP_ERROR DeviceCommissioner::Init(CommissionerInitParams params)
 #endif
 
     mPairingDelegate = params.pairingDelegate;
+    if (params.defaultCommissioner != nullptr)
+    {
+        mDefaultCommissioner = params.defaultCommissioner;
+    }
+    else
+    {
+        mDefaultCommissioner = &mAutoCommissioner;
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:693]: Why still keep an mAutoCommissioner? Why not update the samples to construct an AutoCommissioner and pass it in? Then we wouldn't have to call this `mDefaultCommissioner` but rather `mComissionerDriver` or such

File: src/controller/CHIPDeviceController.h [hunk 2]
```diff
@@ -831,7 +832,10 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
     Callback::Callback<OnNOCChainGeneration> mDeviceNOCChainCallback;
     SetUpCodePairer mSetUpCodePairer;
     AutoCommissioner mAutoCommissioner;
-    CommissioningDelegate * mCommissioningDelegate = nullptr;
+    CommissioningDelegate * mDefaultCommissioner =
+        &mAutoCommissioner; // Commissioning delegate to call when PairDevice / Commission functions are used
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.h:836]: This `= &mAutoCommissioner` overrides the logic in `Init`. It should defualt to null.

File: src/controller/python/ChipDeviceController-ScriptBinding.cpp [hunk 3]
```diff
@@ -92,6 +93,22 @@ chip::Platform::ScopedMemoryBuffer<uint8_t> sSsidBuf;
 chip::Platform::ScopedMemoryBuffer<uint8_t> sCredsBuf;
 chip::Platform::ScopedMemoryBuffer<uint8_t> sThreadBuf;
 chip::Controller::CommissioningParameters sCommissioningParameters;
+
+bool sTestCommissionerUsed = false;
```
Comment (tcarmelveilleux) [inline src/controller/python/ChipDeviceController-ScriptBinding.cpp:97]: Suggest `g` prefix for global. It's more common, and this is effectively a global variable even if static to this module.

File: src/controller/python/test/test_scripts/commissioning_test.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                                   setuppin=20202021,
+                                   nodeid=1),
+              "Failed to finish key exchange")
+
+    logger.info("Testing on off cluster")
+    FailIfNot(test.TestOnOffCluster(nodeid=1,
+                                    endpoint=LIGHTING_ENDPOINT_ID,
+                                    group=GROUP_ID), "Failed to test on off cluster")
+    
+    FailIfNot(test.TestUsedTestCommssioner(), "Test commissioner check failed")
```
Comment (tcarmelveilleux) [inline src/controller/python/test/test_scripts/commissioning_test.py:97]: Typo in `Commissioner`

File: src/controller/CHIPDeviceController.cpp (hunk 1 above)
Comment (cecille) [reply src/controller/CHIPDeviceController.cpp:693]: That's kind of the direction I'm going, but I just haven't gotten there yet - can this be addressed in a follow up? It ends up being a bit dodgy with the CI.

File: src/controller/CHIPDeviceController.h (hunk 2 above)
Comment (cecille) [reply src/controller/CHIPDeviceController.h]: see next commit.

File: src/controller/python/ChipDeviceController-ScriptBinding.cpp (hunk 3 above)
Comment (cecille) [reply src/controller/python/ChipDeviceController-ScriptBinding.cpp:97]: I just used s here to match the style of the file already. When in Rome....

File: src/controller/python/test/test_scripts/commissioning_test.py (hunk 4 above)
Comment (cecille) [reply src/controller/python/test/test_scripts/commissioning_test.py]: fixed.

File: src/controller/CHIPDeviceController.cpp (hunk 1 above)
Comment (tcarmelveilleux) [reply src/controller/CHIPDeviceController.cpp:693]: Sure

**Review by tcarmelveilleux — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: src/controller/AutoCommissioner.h [hunk 5]
```diff
@@ -28,15 +28,14 @@ class DeviceCommissioner;
...
 {
 public:
-    AutoCommissioner(DeviceCommissioner * commissioner) : mCommissioner(commissioner) {}
-    ~AutoCommissioner();
-    CHIP_ERROR SetCommissioningParameters(const CommissioningParameters & params);
-    void SetOperationalCredentialsDelegate(OperationalCredentialsDelegate * operationalCredentialsDelegate);
+    AutoCommissioner() {}
+    virtual ~AutoCommissioner();
+    virtual CHIP_ERROR SetCommissioningParameters(const CommissioningParameters & params) override;
+    virtual void SetOperationalCredentialsDelegate(OperationalCredentialsDelegate * operationalCredentialsDelegate) override;
```
Comment (bzbarsky-apple) [inline src/controller/AutoCommissioner.h:34]:
```suggestion
    CHIP_ERROR SetCommissioningParameters(const CommissioningParameters & params) override;
    void SetOperationalCredentialsDelegate(OperationalCredentialsDelegate * operationalCredentialsDelegate) override;
```
`override` implies `virtual`, so it's just noise to have both... Same for the other APIs here.

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**